### ISSUE TITLE
When I replace fit_generator with fit, behavior is inconsistent.

### ISSUE THREAD

**Issue by yuto3o** [tensorflow/tensorflow#38032]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow):  Yes
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04):  Windows 10
- TensorFlow installed from (source or
binary): conda - TensorFlow version (use command below): 2.0.0
- Python version: 3.7.6

You can collect some of this information using our environment capture
unknown 2.0.0

**Describe the current behavior**
I train a multi-outputs model with a custom data generator. It runs successfully with the api(model.fit_generator). But when I swap to model.fit, it is broken. 

I found that 'fit' can not handle 'multi outputs in list' (such as `yield x, [y1, y2, ...]`) correctly, but tuple(such as `yield x, (y1, y2, ...)`) is ok.

**Describe the expected behavior**
I think both fit_generator and fit should have a consistent behavior to a same generator.

**Standalone code to reproduce the issue** 

```
import numpy as np
from tensorflow.keras import layers, optimizers, losses, Model, Input

inputs = Input(shape=(10,), name='img_input')

x1 = layers.Dense(5)(inputs)
x2 = layers.Dense(2)(inputs)

model = Model(inputs=inputs,
                    outputs=[x1, x2])

model.compile(
    optimizer=optimizers.Adam(),
    loss=losses.categorical_crossentropy)


img_data = np.random.random_sample(size=(1, 10))
targets_0 = np.random.random_sample(size=(1, 5))
targets_1 = np.random.random_sample(size=(1, 2))

def generator_tuple():

    while True:
        yield img_data, (targets_0, targets_1)


def generator_list():
    while True:
        yield img_data, [targets_0, targets_1]

model.fit_generator(generator_tuple(), steps_per_epoch=1, epochs=3) # ok
model.fit(generator_tuple(), steps_per_epoch=1, epochs=3) # ok
model.fit_generator(generator_list(), steps_per_epoch=1, epochs=3) # ok
model.fit(generator_list(), steps_per_epoch=1, epochs=3) # raise error
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
```
Traceback (most recent call last):
  File "C:/Users/yuyang/Documents/codehub/yolov3/bug.py", line 40, in <module>
    model.fit(generator_list(), steps_per_epoch=1, epochs=3) # ok
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\keras\engine\training.py", line 728, in fit
    use_multiprocessing=use_multiprocessing)
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\keras\engine\training_v2.py", line 224, in fit
    distribution_strategy=strategy)
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\keras\engine\training_v2.py", line 547, in _process_training_inputs
    use_multiprocessing=use_multiprocessing)
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\keras\engine\training_v2.py", line 606, in _process_inputs
    use_multiprocessing=use_multiprocessing)
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\keras\engine\data_adapter.py", line 566, in __init__
    reassemble, nested_dtypes, output_shapes=nested_shape)
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\data\ops\dataset_ops.py", line 540, in from_generator
    output_types, tensor_shape.as_shape, output_shapes)
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\data\util\nest.py", line 471, in map_structure_up_to
    results = [func(*tensors) for tensors in zip(*all_flattened_up_to)]
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\data\util\nest.py", line 471, in <listcomp>
    results = [func(*tensors) for tensors in zip(*all_flattened_up_to)]
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\framework\tensor_shape.py", line 1216, in as_shape
    return TensorShape(shape)
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\framework\tensor_shape.py", line 776, in __init__
    self._dims = [as_dimension(d) for d in dims_iter]
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\framework\tensor_shape.py", line 776, in <listcomp>
    self._dims = [as_dimension(d) for d in dims_iter]
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\framework\tensor_shape.py", line 718, in as_dimension
    return Dimension(value)
  File "C:\Users\yuyang\Miniconda3\envs\tf2\lib\site-packages\tensorflow_core\python\framework\tensor_shape.py", line 193, in __init__
    self._value = int(value)
TypeError: int() argument must be a string, a bytes-like object or a number, not 'tuple'
```

**Comment by amahendrakar** [comment]
@yuto3o,
I was able to reproduce the error with [TF v2.0](https://colab.research.google.com/gist/amahendrakar/6845d9d75de52e4bcfd057cdbb930cf3/38032.ipynb). However, the issue seems to be fixed in the latest [TF-nightly](https://colab.research.google.com/gist/amahendrakar/7c1cbc3d1cc01db03b77e2712e77d6dc/38032-tf-nightly.ipynb) version, I was able to run the code without any issues. Please find the attached gist. Thanks!

**Comment by yuto3o** [comment]
Thank you. However, the issue seems to be fixed in TF v2.1.0
